librarian/data/meta/processors.py

```diff
@@ -0,0 +1,411 @@
+            dest_section.update(src_section)
+            # put back (or insert) merged ``dest_section`` that was just
+            # updated with ``src_section`` into ``desctination``
+            destination[language] = dest_section
+
+    def get_path(self):
+        """
+        Return the path with which the extracted metadata should be associated.
+        Overriding this method allows the implementor to redirect the found
+        metadata to another path.
+        """
+        return self.path
+
+    def get_metadata(self):
+        """
+        Return a dict object containing the extracted metadata from
+        py:attr:`~Processor.path`. If the py:attr:`~Processor.partial`` flag
+        is set, or no py:attr:`~Processor.metadata_class` was specified, no
+        metadata extraction will happen.
+        """
+        if self.partial or not self.metadata_extractor:
+            # no additional meta information will be available (besides the
+            # common data)
+            return {}
+        # perform full (possibly expensive) processing of metadata
+        try:
+            meta = self.metadata_extractor.extract()
+        except MetadataError:
+            return {}
+        else:
+            return dict((k, v) for (k, v) in meta.items() if k in self.keys)
+
+    def _add_metadata(self, dest):
+        """
+        Get metadata and merge it into ``dest``.
+        """
+        new_metadata = self.get_metadata()
+        # if metadata is not multilang, wrap it in a dict under ``NO_LANGUAGE``
+        # key, otherwise leave it as-is
+        if not self.metadata_class or not self.metadata_class.multilang:
+            new_metadata = {NO_LANGUAGE: new_metadata}
+        # merge extracted metadata with existing metadata, not overwrite it
+        existing_metadata = dest.get('metadata', {})
+        self._merge(new_metadata, existing_metadata)
+        dest['metadata'] = existing_metadata
+
+    def _add_fs_data(self, dest):
+        """
+        Add generic information that can be gathered without metadata
+        extraction to ``dest``.
+        """
+        # update content types
+        bitmask = ContentTypes.to_bitmask(self.name)
+        content_types = dest.get('content_types', 0) | bitmask
+        # put back updated / merged data
+
+        (mime_type, _) = mimetypes.guess_type(self.path)
+        dest.update(path=self.get_path(),
+                    mime_type=mime_type,
+                    content_types=content_types)
+        # because it is an expensive operation, it is performed only in for
+        # non-partial requests.
+        if not self.partial:
+            # the check should be performed against the original path, not the
+            # path of meta redirection
+            fs_type = (FILE_TYPE, DIRECTORY_TYPE)[self.fsal.isdir(self.path)]
+            dest.update(type=fs_type)
+
+    def process(self):
+        """
+        Process the file system object specified by py:attr:`~Processor.path`.
```

Comment:
This seems to be accurate, because Processor.path will be examined, get_path is used only for redirection.